src/actions/languages/python/motion.ts

```diff
@@ -0,0 +1,256 @@
+    return (
+      elementPosition.line < position.line ||
+      (elementPosition.line === position.line && elementPosition.character < position.character)
+    );
+  }
+
+  /*
+   * Find the position of the specified:
+   *    type: function or class
+   *    direction: next or prev
+   *    edge: start or end
+   *
+   * With this information one can determine all of the required motions
+   */
+  find(type: Type, direction: Direction, edge: Edge, position: Position): Position | null {
+    // Choose the filtering utility based on direction
+    const isDirection = direction === 'next' ? PythonDocument.isAhead : PythonDocument.isBehind;
+
+    // Filter function for all elements whose "edge" is in the correct "direction"
+    // relative to the cursor's position
+    const dir = (element: StructureElement) => isDirection(position, element[edge]);
+
+    // Filter out elements from structure based on type and direction
+    const elements = this.structure.filter((elem) => elem.type === type).filter(dir);
+
+    if (edge === 'end') {
+      // When moving to an 'end' the elements should be started by the end position
+      elements.sort((a, b) => a.end.line - b.end.line);
+    }
+
+    // Return the first match if any exist
+    if (elements.length) {
+      // If direction === 'next' return the first element
+      // otherwise return the last element
+      const index = direction === 'next' ? 0 : elements.length - 1;
+      const element = elements[index];
+      const pos = element[edge];
+
+      // execAction MUST return a fully realized Position object created using new
+      return new Position(pos.line, pos.character);
+    }
+
+    return null;
+  }
+
+  // Use PythonDocument instance to move to specified class boundary
+  static moveClassBoundary(
+    document: TextDocument,
+    position: Position,
+    forward: boolean,
+    start: boolean
+  ): Position {
+    const direction = forward ? 'next' : 'prev';
+    const edge = start ? 'start' : 'end';
+
+    return new PythonDocument(document).find('class', direction, edge, position) || position;
+  }
+}
+
+type Type = 'function' | 'class';
+type Edge = 'start' | 'end';
+type Direction = 'next' | 'prev';
+
+// Uses the specified findFunction to execute the motion coupled to the shortcut (keys)
+abstract class BasePythonMovement extends BaseMovement {
+  type: Type;
+  direction: Direction;
+  edge: Edge;
+
+  public async execAction(position: Position, vimState: VimState): Promise<Position> {
+    const document = vimState.document;
+    switch (document.languageId) {
+      case 'python':
+        return (
+          new PythonDocument(document).find(this.type, this.direction, this.edge, position) ||
+          position
+        );
+
+      default:
+        return position;
+    }
+  }
+}
+
+@RegisterAction
+class MovePythonNextFunctionStart extends BasePythonMovement {
+  keys = [']', 'm'];
+  type = 'function' as Type;
```

Comment:
Are these type assertions necessary?

Reply:
Build fails without this.

Reply:
Ah - making the type explicit while not asserting it should work as well, without sacrificing type safety:

`type: Type = 'function';`

Reply:
Do you know why the type has to be made explicit even though it has been defined in the parent class?

Reply:
Done

Reply:
No I'm not exactly sure why it's necessary, I've just run into similar things before

Reply:
This should be `type: Type = 'function';`, as it's more type-safe

Reply:
Oh sorry, is this fixed? GitHub's UI is confusing me. The same improvement should be made to `direction` and `edge`, if it's not already

Reply:
Done